pkg/elb/cleanup.go

```diff
@@ -0,0 +1,299 @@
+	for _, tag := range tags {
+		if aws.StringValue(tag.Key) == clusterTagKey {
+			return true
+		}
+	}
+	return false
+}
+
+func getSecurityGroupsOwnedByLoadBalancer(ctx context.Context, ec2API ec2iface.EC2API, elbAPI elbiface.ELBAPI,
+	clusterName string, loadBalancerName string, loadBalancerKind int) (map[string]struct{}, error) {
+	if loadBalancerKind == loadBalancerKindNetwork {
+		// V2 ELBs just use the Security Group of the EC2 instances
+		return map[string]struct{}{}, nil
+	}
+
+	lb, err := describeClassicLoadBalancer(ctx, elbAPI, loadBalancerName)
+	if err != nil {
+		return nil, err
+	}
+	if lb == nil {
+		// The load balancer wasn't found
+		return map[string]struct{}{}, nil
+	}
+
+	sgResponse, err := describeSecurityGroupsByID(ctx, ec2API, aws.StringValueSlice(lb.SecurityGroups))
+
+	if err != nil {
+		return nil, fmt.Errorf("error querying security groups for ELB %s: %s", loadBalancerName, err)
+	}
+
+	result := map[string]struct{}{}
+
+	for _, sg := range sgResponse.SecurityGroups {
+		sgID := aws.StringValue(sg.GroupId)
+
+		// FIXME(fons): AWS' CloudConfig  accepts a global ELB security group, which shouldn't be deleted.
+		//              However, there doesn't seem to be a way to access the CloudConfiguration through the API Server.
+		// if sgID == cfg.Global.ElbSecurityGroup {
+		//	//We don't want to delete a security group that was defined in the Cloud Configuration.
+		//	continue
+		// }
+
+		if sgID == "" {
+			continue
+		}
+
+		// Only delete security groups created by the Kubernetes Cloud provider
+		if !tagsIncludeClusterName(sg.Tags, clusterName) {
+			continue
+		}
+
+		result[sgID] = struct{}{}
+	}
+
+	return result, nil
+}
+
+func getLoadBalancerKind(service *corev1.Service) int {
+	// See https://github.com/kubernetes/kubernetes/blob/v1.12.6/pkg/cloudprovider/providers/aws/aws_loadbalancer.go#L51-L56
+	if service.Annotations[awsprovider.ServiceAnnotationLoadBalancerType] == "nlb" {
+		return loadBalancerKindNetwork
+	}
+	return loadBalancerKindClassic
+}
+
+func loadBalancerExists(ctx context.Context, elbAPI elbiface.ELBAPI, elbv2API elbv2iface.ELBV2API,
+	name string, kind int) (bool, error) {
+	if kind == loadBalancerKindNetwork {
+		return elbV2Exists(ctx, elbv2API, name)
+	}
+	desc, err := describeClassicLoadBalancer(ctx, elbAPI, name)
+	return desc != nil, err
+}
+
+func describeClassicLoadBalancer(ctx context.Context, elbAPI elbiface.ELBAPI,
+	name string) (*elb.LoadBalancerDescription, error) {
+
+	request := &elb.DescribeLoadBalancersInput{
+		LoadBalancerNames: []*string{&name},
+	}
+
+	response, err := elbAPI.DescribeLoadBalancersWithContext(ctx, request)
+	if err != nil {
+		if awsError, ok := err.(awserr.Error); ok {
+			if awsError.Code() == "LoadBalancerNotFound" {
+				return nil, nil
+			}
+		}
+		return nil, err
+	}
+
+	var ret *elb.LoadBalancerDescription
+	for _, loadBalancer := range response.LoadBalancerDescriptions {
```

Comment:
why not just check the size of the array for the warning?

Reply:
Yeah, that makes more sense. I just copy-pasted from https://github.com/kubernetes/legacy-cloud-providers/blob/release-1.15/aws/aws.go#L2728-L2734